### PR TITLE
Fix TraceableTcpTransportChannel classcast exception

### LINKED ISSUE #3515: [BUG]TraceableTcpTransportChannel class cast exception,
**What is the bug?**
When the experimental feature flag (opensearch.experimental.feature.telemetry.enabled) is set to true and the security plugin installed. There is a bootstrap error.

Caused by: org.opensearch.core.common.io.stream.NotSerializableExceptionWrapper: class_cast_exception: class org.opensearch.telemetry.tracing.channels.TraceableTcpTransportChannel cannot be cast to class org.opensearch.transport.TcpTransportChannel (org.opensearch.telemetry.tracing.channels.TraceableTcpTransportChannel and org.opensearch.transport.TcpTransportChannel are in unnamed module of loader 'app')
	at org.opensearch.security.ssl.transport.SecuritySSLRequestHandler.messageReceived(SecuritySSLRequestHandler.java:116) ~[?:?]
	at org.opensearch.security.OpenSearchSecurityPlugin$6$1.messageReceived(OpenSearchSecurityPlugin.java:790) ~[?:?]
	at org.opensearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:106) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundHandler.handleRequest(InboundHandler.java:262) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundHandler.messageReceived(InboundHandler.java:140) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundHandler.inboundMessage(InboundHandler.java:123) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.transport.TcpTransport.inboundMessage(TcpTransport.java:770) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:175) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:150) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:115) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:95) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[?:?]
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:280) ~[?:?]

**How can one reproduce the bug?**

**What is the expected behavior?**

**What is your host/environment?**
 - Plugins

**Do you have any screenshots?**

**Do you have any additional context?**
Add any other context about the problem.

### PR DESCRIPTION
**Pull request by Gaganjuneja**
### Description
Fix the TraceableTcpTransportChannel class cast exception when enabling the feature opensearch.experimental.feature.telemetry.enabled. Issue reported here - https://github.com/opensearch-project/security/issues/3464#issuecomment-1753641505
### Issues Resolved
#3515

Reference Chnage - https://github.com/opensearch-project/OpenSearch/blob/8e5e54e87f6f86a2425a5c767e208767ba3df14c/server/src/main/java/org/opensearch/transport/RequestHandlerRegistry.java#L101

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
[Please provide details of testing done: unit testing, integration testing and manual testing]

### Check List
- [x] New functionality includes testing
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by Gaganjuneja** [comment]
@reta, @peternied Please take a look at it. We need to merge this fix in 2.11

File: src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java [hunk 1]
```diff
@@ -113,9 +113,9 @@ public final void messageReceived(T request, TransportChannel channel, Task task
 
             if (channel instanceof TaskTransportChannel) {
                 final TransportChannel inner = ((TaskTransportChannel) channel).getChannel();
-                nettyChannel = (Netty4TcpChannel) ((TcpTransportChannel) inner).getChannel();
-            } else if (channel instanceof TcpTransportChannel) {
-                final TcpChannel inner = ((TcpTransportChannel) channel).getChannel();
+                nettyChannel = (Netty4TcpChannel) ((BaseTcpTransportChannel) inner).getChannel();
+            } else if (channel instanceof BaseTcpTransportChannel) {
+                final TcpChannel inner = ((BaseTcpTransportChannel) channel).getChannel();
```
Comment (peternied) [inline src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java:118]: Lets add a test(s) that confirms this fix is resolved.

If you are looking for an example of how these tests could be written, take a look at https://github.com/opensearch-project/security/blob/main/src/integrationTest/java/org/opensearch/security/http/OnBehalfOfJwtAuthenticationTest.java that might provide some guidance on cluster configuration and a basic setup.

**Review by peternied — CHANGES_REQUESTED** [review]
@Gaganjuneja Thanks for creating this PR - if this is breaking it seems like we are missing some tests for these scenarios.

**Comment by Gaganjuneja** [comment]
> @Gaganjuneja There is another issue with the Telemetry experimental feature and the security plugin. The way that the TraceableHttpChannel wraps the underlying channel [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/http/AbstractHttpServerTransport.java#L367) is problematic. Security needed to recently refactor where auth was performed from the last step of the netty pipeline to further up the pipeline and needs access to the netty channel.
> 
> You can see an example here of how the security plugin would access the channel: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityRestFilter.java#L135

@reta, This seems to be new changes which will be bypassed for TraceableHttpChannel always. I am thinking, if we can add a new method to the HttpChannel::getType and let all the implementations implement it.

```
default String getType() {
    return ChannelType.DEFAULT;
}
```

For Netty4

```
public String getType() {
    return ChannelType.NETTY4;
}
```

**Comment by reta** [comment]
> @reta, This seems to be new changes which will be bypassed for TraceableHttpChannel always. I am thinking, if we can add a new method to the HttpChannel::getType and let all the implementations implement it.

@Gaganjuneja  I sadly cannot look at the issue today, travelling back home, but it does not look right - the HTTP transport should have no dependency on the engine behind it, if any, I will take a look tomorrow morning

**Comment by Gaganjuneja** [comment]
> > @reta, This seems to be new changes which will be bypassed for TraceableHttpChannel always. I am thinking, if we can add a new method to the HttpChannel::getType and let all the implementations implement it.
> 
> @Gaganjuneja I sadly cannot look at the issue today, travelling back home, but it does not look right - the HTTP transport should have no dependency on the engine behind it, if any, I will take a look tomorrow morning

Sure, sorry for bothering you!

File: src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java (hunk 1 above)
Comment (Gaganjuneja) [reply src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java:118]: @peternied, I have written a Integ test case and its failing for 2 more issues. Where RestRequest::getHttpChannel is being type casted to Netty4HttpChannel. At one place this is already been reported by @cwperks and in addition to that one check is here https://github.com/opensearch-project/security/blob/e69013651189238f3c93d57415d37134c8bfd6df/src/main/java/org/opensearch/security/filter/OpenSearchRequest.java#L47

**Comment by Gaganjuneja** [comment]
> > @reta, This seems to be new changes which will be bypassed for TraceableHttpChannel always. I am thinking, if we can add a new method to the HttpChannel::getType and let all the implementations implement it.
> 
> @Gaganjuneja I sadly cannot look at the issue today, travelling back home, but it does not look right - the HTTP transport should have no dependency on the engine behind it, if any, I will take a look tomorrow morning

Other simplest option could be is to have TraceableHttpChannel::getDelegateChannel method and inside the security plugin we can put an additional check 

```
HttpChannel channel = restRequest.getChannel();
if(channel instanceof TraceableHttpChannel){
   channel = ((TraceableHttpChannel)channel).getDelegateChannel();
}

//Now the Netty4 check could proceed.
```

File: src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java (hunk 1 above)
Comment (reta) [reply src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java:118]: @Gaganjuneja @peternied I see at least 2 issues with this code:
 - it very coupled with Netty, that should not be the case
 - the issue will come back when we would instrument `TaskTransportChannel`

Would I would like to suggest: add `get` (and possibly `put`, have to look into it) methods to `HttpChannel` and `TcpChannel`:

```
public T get(String attribute, Class<T> clazz);
```

With ytat, we could remove `|| !(underlyingRequest.getHttpChannel() instanceof Netty4HttpChannel)) {` from `OpenSearchRequest` since it will be `null` value anyway.

Comment (Gaganjuneja) [reply src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java:118]: Thanks @reta, Yes it would be more cleaner approach.

Comment (Gaganjuneja) [reply src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java:118]: There is one more change where we are trying to get the SSLHandler from the inbound and outbound pipeline

https://github.com/opensearch-project/security/blob/e69013651189238f3c93d57415d37134c8bfd6df/src/main/java/org/opensearch/security/filter/OpenSearchRequest.java#L54

Comment (reta) [reply src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java:118]: Ir does not matter, it is all covered, will send the pull request shortly

Comment (reta) [reply src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java:118]: https://github.com/opensearch-project/OpenSearch/pull/10562

Comment (Gaganjuneja) [reply src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java:118]: One clarification, Please pardon me for the naive question. How will we access the channel attributes (NettyAttribute::proFrom) for authentication? Or the old implementation should be suffice.

Comment (reta) [reply src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java:118]: > One clarification, Please pardon me for the naive question. How will we access the channel attributes (NettyAttribute::popFrom) for authentication? Or the old implementation should be suffice.

@Gaganjuneja chatted with Craig (@cwperks) offline, it seems like we may try to pursue a different path for early termination of the the request, I would suggest we do nothing for channel attributes (NettyAttribute::*) now since it is likely to go away (sadly no chances to fix that `2.11`). It should not be an issue, tracing is an experimental feature that is expected to have hiccups.

Comment (reta) [reply src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java]: @Gaganjuneja I've updated the change with respect to core changes

**Review by cwperks — APPROVED** [review]

**Review by willyborankin — APPROVED** [review]

**Review by peternied — COMMENTED** [review]
Thanks - this change looks much cleaner.  Happy to re-review after an integration test has been added.

**Comment by reta** [comment]
> Thanks - this change looks much cleaner. Happy to re-review after an integration test has been added.

@Gaganjuneja are you working on that?

**Comment by Gaganjuneja** [comment]
> > Thanks - this change looks much cleaner. Happy to re-review after an integration test has been added.
> 
> @Gaganjuneja are you working on that?

@reta, I will pick this up very first thing next week.

**Comment by Gaganjuneja** [comment]
> > > Thanks - this change looks much cleaner. Happy to re-review after an integration test has been added.
> > 
> > 
> > @Gaganjuneja are you working on that?
> 
> @reta, I will pick this up very first thing next week.

There is nothing we can test about telemetry. Only thing is to enable it and see if cluster is coming up fine as far as the integration with security plugin is concerned. I have enable it for SearchOperationTest.java and now it should be able to catch if anything is breaking due to change in the Telemetry side.

File: src/integrationTest/java/org/opensearch/security/SearchOperationTest.java [hunk 2]
```diff
@@ -371,6 +373,16 @@ public class SearchOperationTest {
             USER_ALLOWED_TO_PERFORM_INDEX_OPERATIONS_ON_SELECTED_INDICES,
             USER_ALLOWED_TO_CREATE_INDEX
         )
+        .nodeSettings(
```
Comment (reta) [inline src/integrationTest/java/org/opensearch/security/SearchOperationTest.java:376]: This flag is noop if plugin is not installed, right? I think what we could do is to enhance (or add new) the GA which does full plugin install test to have security and opentelemetry plugin installed together, @peternied does it sound right to you?

Comment (Gaganjuneja) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java:376]: yes @reta you are right. but this would atl east create the Traceable wappers.

Comment (reta) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java:376]: > yes @reta you are right. but this would at least create the Traceable wrappers.

how? it will use NOOP tracer

Comment (Gaganjuneja) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java:376]: Oh, yes. After this recent change tracer:isTracingEnabled, it wont.

File: src/integrationTest/java/org/opensearch/security/SearchOperationTest.java [hunk 3]
```diff
@@ -371,6 +373,9 @@ public class SearchOperationTest {
...
+            Map.of(FeatureFlags.TELEMETRY_SETTING.getKey(), true, TelemetrySettings.TRACER_FEATURE_ENABLED_SETTING.getKey(), true)
```
[the remark this answers is not in the document]
Comment (Gaganjuneja) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java]: @peternied, As mentioned by @reta these settings are not going to work and tracing would be noop in this scenario. In case of NoopTracer we tried to implement in a way that its shouldn't alter the execution pattern by unnecessarily introducing the tracer related core changes (Tracing being an experimental feature eg. is TraceableHttpChannel here. If tracer is noop we are not wrapping the HttpChannel hence its working with FeatureFlag disabled).

Now to test the telemetry plugin and security plugin we need to install them together and execute the tests.

To validate the telemetry working we need to have some search/indexing requests and validate if traces are being generated and which is not very straight forward.

[the remark this answers is not in the document]
Comment (peternied) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java]: >To validate the telemetry working we need to have some search/indexing requests and validate if traces are being generated and which is not very straight forward.

[the remark this answers is not in the document]
Comment (Gaganjuneja) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java]: How are we testing security with other plugins today?

[the remark this answers is not in the document]
Comment (Gaganjuneja) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java]: One more thing, Even if we write the integration test, it's still going to fail bcauase the issue is partially fixed and the check inside NettyAttribute.java is still there - 

```
public static <T> Optional<T> popFrom(final RestRequest request, final AttributeKey<T> attribute) {
        if (request.getHttpChannel() instanceof Netty4HttpChannel) {
            Channel nettyChannel = ((Netty4HttpChannel) request.getHttpChannel()).getNettyChannel();
            return Optional.ofNullable(nettyChannel.attr(attribute).getAndSet(null));
        }
        return Optional.empty();
    }
```

[the remark this answers is not in the document]
Comment (reta) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java]: Ah, correct, this part is on me, thanks for reminding @Gaganjuneja

[the remark this answers is not in the document]
Comment (cwperks) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java]: @Gaganjuneja You can see an example of how anomaly-detection tests with security here: https://github.com/opensearch-project/anomaly-detection/blob/main/.github/workflows/test_security.yml

They run their full integ test suite in a cluster w/ the security plugin installed.

File: src/integrationTest/java/org/opensearch/security/SearchOperationTest.java (hunk 2 above)
Comment (peternied) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java]: I've moved this flag to its own class - but I was not able to reproduce a failure with the other changes in this CR reverted.  

Please pull the latest changes from your fork and test locally until the test can fail if you'd reverted the other changes - its possible I've got the feature flag configured incorrectly.

File: src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java [hunk 4]
```diff
@@ -108,21 +104,7 @@ public final void messageReceived(T request, TransportChannel channel, Task task
         }
 
         try {
-
-            Netty4TcpChannel nettyChannel = null;
-
```
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java:113]: curious to understand why these changes were made?

Comment (reta) [reply src/main/java/org/opensearch/security/ssl/transport/SecuritySSLRequestHandler.java:113]: To get rid of type casting

**Comment by DarshitChanpura** [comment]
@Gaganjuneja would you need any helping getting this across?

**Comment by DarshitChanpura** [comment]
@Gaganjuneja are you still working on this?

File: src/integrationTest/java/org/opensearch/security/SearchOperationTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gaganjuneja) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java]: > One more thing, Even if we write the integration test, it's still going to fail bcauase the issue is partially fixed and the check inside NettyAttribute.java is still there -
> 
> ```
> public static <T> Optional<T> popFrom(final RestRequest request, final AttributeKey<T> attribute) {
>         if (request.getHttpChannel() instanceof Netty4HttpChannel) {
>             Channel nettyChannel = ((Netty4HttpChannel) request.getHttpChannel()).getNettyChannel();
>             return Optional.ofNullable(nettyChannel.attr(attribute).getAndSet(null));
>         }
>         return Optional.empty();
>     }
> ```



> Ah, correct, this part is on me, thanks for reminding @Gaganjuneja

@reta, What is your opinion, this should be addressed here on in another PR. 

@DarshitChanpura, Once the above issue is addressed, I will write the integration test.

[the remark this answers is not in the document]
Comment (reta) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java]: > @reta, What is your opinion, this should be addressed here on in another PR.

Up to you @Gaganjuneja

**Comment by peternied** [comment]
Doesn't look like this is being worked on anymore, going to close this out. Please feel free to reopen if you re-engage on this @Gaganjuneja
